File: app/src/main/java/james/metronome/views/MetronomeView.java
package james.metronome.views;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.DecelerateInterpolator;

import com.afollestad.aesthetic.Aesthetic;

import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

public class MetronomeView extends View {

    private Paint paint;
    private Paint accentPaint;
    private long interval = 500;
    private float distance;
    private boolean isEmphasis;

    private Disposable colorAccentSubscription;
    private Disposable textColorPrimarySubscription;

    public MetronomeView(Context context) {
        this(context, null);
    }

    public MetronomeView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MetronomeView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        paint = new Paint();
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeWidth(2);
        paint.setAntiAlias(true);
        paint.setColor(Color.BLACK);

        accentPaint = new Paint();
        accentPaint.setStyle(Paint.Style.STROKE);
        accentPaint.setStrokeWidth(8);
        accentPaint.setAntiAlias(true);
        accentPaint.setColor(Color.BLACK);

        subscribe();
    }

    public void subscribe() {
        colorAccentSubscription = Aesthetic.Companion.get()
                .colorAccent()
                .subscribe(new Consumer<Integer>() {
                    @Override
                    public void accept(Integer integer) throws Exception {
                        accentPaint.setColor(integer);
                    }
                });

        textColorPrimarySubscription = Aesthetic.Companion.get()
                .textColorPrimary()
                .subscribe(new Consumer<Integer>() {
                    @Override
                    public void accept(@NonNull Integer integer) {
                        paint.setColor(integer);
                    }
                });
    }

    public void unsubscribe() {
        colorAccentSubscription.dispose();
        textColorPrimarySubscription.dispose();
    }

    public void setInterval(long interval) {
        this.interval = interval;
    }

    public void onTick(boolean isEmphasis) {
        this.isEmphasis = isEmphasis;

        ValueAnimator animator = ValueAnimator.ofFloat(0, 1);
        animator.setDuration(interval);
        animator.setInterpolator(new DecelerateInterpolator());
        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                distance = (float) animation.getAnimatedValue();
                invalidate();
            }
        });
        animator.start();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        paint.setAlpha((int) (255 * (1 - distance)));
        accentPaint.setAlpha((int) (255 * (1 - distance)));
        canvas.drawCircle(canvas.getWidth() / 2, canvas.getHeight() / 2, distance * Math.max(canvas.getWidth(), canvas.getHeight()) / 2, isEmphasis ? accentPaint : paint);
        //this probably draws a circle or something idk
    }
}

File: app/src/main/java/me/jfenn/metronome/views/MetronomeView.kt
package me.jfenn.metronome.views

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import android.view.animation.DecelerateInterpolator
import me.jfenn.metronome.R
import me.jfenn.metronome.utils.getThemedColor

class MetronomeView @JvmOverloads constructor(
        context: Context?,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private val paint = Paint().apply {
        style = Paint.Style.STROKE
        strokeWidth = 2f
        isAntiAlias = true
        color = context?.getThemedColor(android.R.attr.textColorPrimary) ?: Color.BLACK
    }

    private val accentPaint = Paint().apply {
        style = Paint.Style.STROKE
        strokeWidth = 8f
        isAntiAlias = true
        color = context?.getThemedColor(R.attr.colorAccent) ?: Color.RED
    }

    private var interval: Long = 500
    private var distance = 0f
    private var isEmphasis = false

    fun setInterval(interval: Long) {
        this.interval = interval
    }

    fun onTick(isEmphasis: Boolean) {
        this.isEmphasis = isEmphasis
        val animator = ValueAnimator.ofFloat(0f, 1f)
        animator.duration = interval
        animator.interpolator = DecelerateInterpolator()
        animator.addUpdateListener { animation ->
            distance = animation.animatedValue as Float
            invalidate()
        }
        animator.start()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        paint.alpha = (255 * (1 - distance)).toInt()
        accentPaint.alpha = (255 * (1 - distance)).toInt()
        canvas.drawCircle(canvas.width / 2.toFloat(), canvas.height / 2.toFloat(), distance * Math.max(canvas.width, canvas.height) / 2, if (isEmphasis) accentPaint else paint)
        //this probably draws a circle or something idk
    }
}